### PR TITLE
(duplicate) Zim-Wiki support nr1 pschwede

### PR DESCRIPTION
**Pull request by pschwede**
ref  Format: Zim Wiki #393 

---------

Hi. Sorry, my PR is premature. I asked @gsantner some question concerning getting my `ZimWikiTextConverter` on Matrix. In order answer it he needed some code to see and suggested to post a PR.

#### Issues
- [x] Links to other files like `[[PageLink]]` and `[[:NotebookLink]]` cause line to be completely invisible.
- [x] Syntax highlighting doesn't work for `== headings ==`.
- [x] Syntax highlighting doesn't work for `[[Links]]` (other than `[[web|links]]`).
- [ ] support `{{images.jpg}}`
- [ ] `__marked__` has to get a yellow background color (or similar according to theme)
- [ ] support tables
- [ ] support check boxes `[>]` and `[*]`
- [ ] support `'''Verbatim block'''`
- [x] support `_{Subscript}`
- [x] support `^{Superscript}`
- [x] ~~:question: make check boxes interactive in HTML view~~
- [ ] Turn ZimWikiTextActions into useful ones
   - [ ] Insert Zim Wiki Header action
- [ ] Take a look into ZimWikiAutoFormat
- [ ] (Optional) support `$${TeX}$$`

### CODE REVIEW

**Comment by gsantner** [comment]
@fredericjacob you said to also have started on it right? Is there some duplicated work already? Guess we will result in one "base" PR that adds zim in general, and every others add on top (we can split into multiple smaller ones if required/helpful)

**Comment by gsantner** [comment]
this is what you need to take a look at. Add .zim to TextFormat.CONVERTER_ZIM.isFileOutOfThisFormat

grafik

**Comment by pschwede** [comment]
Thanks for taking a look.

The file extension of a zim file is `.txt`, not `.zim`. Please do not confuse [zim-wiki](https://zimwiki.org) with [openzim](https://openzim.org/).

**Comment by gsantner** [comment]
Oh, then please do not assign the extension to default to zim. txt should stay with plaintext for new opened documents. However for development you can change it for the time being.

File: app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiAutoFormat.java [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+            final String checkString = uLine.isCheckboxList ? checkSymbol : "";
+            result = indent + String.format("%c %s", uLine.listChar, checkString);
+        } else {
+            result = indent;
+        }
+
+        return result;
+    }
+
+    public static class ListLine {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiAutoFormat.java:59]: We should try not to duplicate this code. If it is applicable to both zim as well as markdown, we should pull this code into a separate module and import it as appropriate.

Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiAutoFormat.java:59]: I think it just started off as copy-paste from markdown code

Comment (pschwede) [reply app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiAutoFormat.java:59]: I'll take a look. Thanks!

**Review by fredericjacob — COMMENTED** [review]
@pschwede Hi, I reviewed your code and made some comments, in particular regarding the merge with #1088 so that we get a single code base to build upon.

The biggest thing that I found: `ListHandler`, `ZimWikiHighlighter` (of which I have a working implementation in #1088), `ZimWikiTextActions`, `ZimWikiAutoFormat` are right now only very slightly changed copies of their markdown counterparts, and I don't see why they're needed. By now, it's just duplicated code which clutters the project unneccessarily.
If a `TextAction` or `AutoFormat` object is needed somewhere, we can just use the markdown class for now. Later, we can replace this with Zim-specific implementations. If these need the same functionality as the markdown classes, we should extract part of it to a superclass or some module and reuse it for both variants.

So I suggest to delete these four classes (and replace potential usages with their markdown counterparts) before we merge our PRs. This makes it easier to build on top of it for both of us and to make clear which features are already there and which aren't.

Most tests are also copy-pasted from the markdown test files to a great extent. I think It would be better to reuse code here too, where applicable, but it's not such a big issue like for the non-test code.

Another thing where we should agree upon before merging are the pattern namings - I made some comments to `ZimWikiHighlightingPattern`. But this is more a minor thing of style and consistency. :)

File: app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiHighlighterPattern.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    LIST_CHECK(Pattern.compile("^\t*(\\[[ xX*>]?\\]|\\([ xX*>]?\\)) ")),
+    LIST_ORDERED(Pattern.compile("^\t*([\\d]+\\.|[a-zA-Z]+\\.) ")),
+    LIST_UNORDERED(Pattern.compile("^\t*(\\*)(?= )")),
+    EMPHASIS(Pattern.compile("(//(?!/)(.*?)(?<!:)//)")),
+    STRONG(Pattern.compile("(\\*\\*(?!\\*)(.*?)\\*\\*)")),
+    MARK(Pattern.compile("(__(?!_)(.*?)__)")),
+    STRIKE(Pattern.compile("(~~(?!~)(.+?)~~)")),
+    SUBSCRIPT(Pattern.compile("(_\\{(?!~)(.+?)\\})")),
+    SUPERSCRIPT(Pattern.compile("(\\^\\{(?!~)(.+?)\\})")),
+    VERBATIM(Pattern.compile("(''(?!').+?'')")),
```
[the remark this answers is not in the document]
Comment (pschwede) [reply app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiHighlighterPattern.java:27]: I'd like to stick with Zim-Wiki internal names.

**Review by pschwede — COMMENTED** [review]
Sorry, my PR is premature. I asked @gsantner some question on Matrix and to answer it he needed some code to see and suggested to post a PR.

**Comment by gsantner** [comment]
Hi you two,

as said, yes, both have rough edges. I dont expect perfect right from start. I leave the coordination of you both up to you. As said, I would go for integrating the highlighter into pschwede first.

File: app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiAutoFormat.java (hunk 1 above)
Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiAutoFormat.java:59]: closing this thread for now, we can take a look again when it's further progressed

**Comment by fredericjacob** [comment]
Hi @pschwede, I merged your recent changes into #1088. I came across some errors in the tests and I also identified mistakes in the regexes I wrote thanks to the testcases. Fixed both of them. On top of that, I added some few more editor highlighting features. So my branch is right now up-to-date with yours and you can fast-forward by merging it into yours, to keep them in sync. :)

@gsantner @pschwede Yes, of course the code does not have to be perfect right from the start. And I really didn't want to rant in any way where I suggested or commented something. I believe that feedback is important (especially as we are cooperating here and working on the same code) and I appreciate any feedback/reviews on my own code as well, which I know is also not perfect.
From my experience, it's just easier to work on a shared code base (which is in most cases the master branch) if it is more or less "clean" and free of obvious errors, with the tests passing. If it's a personal branch, it's not an issue at all how premature the code is. Then it's enough to clean up just before pushing to the shared branch.
I think this is what makes the coordination in our case a bit difficult, as this PR is in fact linked to pschwedes personal branch. 

However, so far it worked out for now - maybe I was also thinking a bit too critical. The PRs are (almost) merged together, editor highlighting is there and the view mode is also mostly working from what I saw.

@gsantner So how should we best continue from here? Are you going to review our combined code?

As I am done with the editor highlighter for now, I'm looking forward to implement some text actions. Probably it makes sense to create a new PR for every action (set) I'm going to implement? But on which code base should I build upon for this? On this PR? Or first wait until it is merged into the master and then start from there?
If the current code still needs some time until maturity - maybe it makes sense to create a separate branch for @pschwede and me which we can use as a basis for zim development?

**Comment by gsantner** [comment]
@pschwede 
Please see #1098

OK with that?

**Pull request closed without merging**